src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -2837,11 +2837,26 @@ std::shared_ptr<primitive_impl> ImplementationsFactory::get_primitive_impl_for_p
         o.data_padding._dynamic_dims_mask = padding::EMPTY_MASK;
     }
 
+    auto need_single_batch_optimization = [&inst, &updated_params](const std::shared_ptr<primitive_impl> impl) -> bool {
+        auto is_cldnn_fc_impl = inst.get_node().get_preferred_impl_type() == impl_types::ocl;
+        auto kernel_name = impl->get_kernel_name();
+        // Avoid ref_kernel test issue.
+        auto is_ref_impl = kernel_name.find("fully_connected_gpu_bfyx_ref") != std::string::npos;
+        auto is_gemv_impl = kernel_name.find("gemv") != std::string::npos;
+        return is_cldnn_fc_impl && fully_connected_inst::can_apply_single_batch_optimization(updated_params) &&
+               !is_ref_impl && !is_gemv_impl;
+    };
+
     // 1. If we have static impl in the cache - use it
     if (use_async_compilation && ((inst.get_impl() && inst.get_impl()->is_dynamic()) || inst.get_flag(ExecutionFlags::SHAPE_CHANGED))) {
         auto cached_impl = m_static_impls_cache.get(updated_params);
         if (cached_impl) {
-            return cached_impl->clone();
+            if (inst.get_node().is_type<fully_connected>() && need_single_batch_optimization(cached_impl)) {
+                // Switch to single batch optimization.
+                cached_impl = nullptr;
+            } else {
+                return cached_impl->clone();
+            }
```

Comment:
Why do we need this custom logic for this particular case? As I can see, async compilation is already enabled specifically for u4/i8 weights with batch_size=1

Reply:
Confirmed, this branch is not needed, thanks!